libnavui-androidauto/build.gradle

```diff
@@ -34,25 +34,35 @@ android {
 }
 
 dependencies {
+    // Includes a version of the Mapbox Android Auto extension. This library also includes the
+    // Jetpack car library.
     api(dependenciesList.mapboxMapsAndroidAuto)
-    api(project(":libnavigation-android"))
-    implementation(project(":libnavui-resources"))
-    implementation(project(":libnavui-util"))
-    api("com.mapbox.search:mapbox-search-android:1.0.0-beta.29")
+
+    // Downstream must specify a version of navigation and search.
+    // Maps, Navigation, and Search must be in sync to include the same Common sdk version.
+    compileOnly(project(":libnavigation-android"))
```

Comment:
Yeah we had a lot of discussion on the maps side to [make android auto an extension](https://github.com/mapbox/mapbox-maps-android/blob/631239afd7053291fd6e50b823aa7a01ecff53fc/extension-androidauto/build.gradle.kts#L25). Also, this library is a little different definition so we may not want to follow the same extension philosophy.  [Here is some discussion](https://github.com/mapbox/mapbox-maps-android/pull/1267#discussion_r845282315). 

There are trade-offs. You're listing some of the cons, obscure dependencies. But are also positives.

- Allows the nav-sdk to be released often without new releases of libnavigation-android
- When we identify compatibility issues, we will keep track of them with new releases of `libnavigation-android`
- This will also allow us to use `libnavigation-android` with older versions of Nav-SDK - technically, this release has been tested with all of nav-sdk:2.*
- The real issues you run into will be at **compile time**. A developer will not be able to compile an app that does not work

Reply:
> developers couldn't just choose a version of Nav SDK, it would always have to be the latest or even a snapshot of unreleased changes. This would be very hard for us to keep in check

Are you sure about this? You should be able to bump the Nav-SDK and get new apis no problem with this set up. That's actually the reason to do this. In 1tap for example, we can bump the nav-sdk and not bump androidauto

Reply:
> How does the POM generated by this configuration look like for the Android Auto artifact?

I was also curious here so just published a new snapshot to ensure this latest version is used.

[ui-androidauto-0.1.0-SNAPSHOT.pom.txt](https://github.com/mapbox/mapbox-navigation-android/files/8636409/ui-androidauto-0.1.0-SNAPSHOT.pom.txt) 
<details closed>
<summary>Click me to see</summary>
<pre>

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <!-- This module was also published with a richer model, Gradle metadata,  -->
  <!-- which should be used instead. Do not delete the following line which  -->
  <!-- is to indicate to Gradle or any Gradle module metadata file consumer  -->
  <!-- that they should prefer consuming it instead. -->
  <!-- do_not_remove: published-with-gradle-metadata -->
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.mapbox.navigation</groupId>
  <artifactId>ui-androidauto</artifactId>
  <version>0.1.0-SNAPSHOT</version>
  <packaging>aar</packaging>
  <dependencies>
    <dependency>
      <groupId>com.mapbox.extension</groupId>
      <artifactId>maps-androidauto</artifactId>
      <version>0.1.0</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.jetbrains.kotlin</groupId>
      <artifactId>kotlin-parcelize-runtime</artifactId>
      <version>1.5.31</version>
      <scope>runtime</scope>
    </dependency>
    <dependency>
      <groupId>androidx.appcompat</groupId>
      <artifactId>appcompat</artifactId>
      <version>1.3.1</version>
      <scope>runtime</scope>
    </dependency>
    <dependency>
      <groupId>org.jetbrains.kotlinx</groupId>
      <artifactId>kotlinx-coroutines-core</artifactId>
      <version>1.5.1</version>
      <scope>runtime</scope>
    </dependency>
    <dependency>
      <groupId>androidx.lifecycle</groupId>
      <artifactId>lifecycle-runtime-ktx</artifactId>
      <version>2.4.0</version>
      <scope>runtime</scope>
    </dependency>
    <dependency>
      <groupId>androidx.lifecycle</groupId>
      <artifactId>lifecycle-livedata-ktx</artifactId>
      <version>2.4.0</version>
      <scope>runtime</scope>
    </dependency>
    <dependency>
      <groupId>androidx.constraintlayout</groupId>
      <artifactId>constraintlayout</artifactId>
      <version>1.1.3</version>
      <scope>runtime</scope>
    </dependency>
    <dependency>
      <groupId>androidx.datastore</groupId>
      <artifactId>datastore-preferences</artifactId>
      <version>1.0.0</version>
      <scope>runtime</scope>
    </dependency>
    <dependency>
      <groupId>org.jetbrains.kotlin</groupId>
      <artifactId>kotlin-stdlib-jdk8</artifactId>
      <version>1.5.31</version>
      <scope>runtime</scope>
    </dependency>
    <dependency>
      <groupId>org.jetbrains.kotlinx</groupId>
      <artifactId>kotlinx-coroutines-android</artifactId>
      <version>1.5.1</version>
      <scope>runtime</scope>
    </dependency>
  </dependencies>
  <name>Android Auto Mapbox Navigation SDK</name>
  <description>Artifact that allows users to use Mapbox Navigation in Android Auto</description>
  <url>https://github.com/mapbox/mapbox-navigation-android</url>
  <licenses>
    <license>
      <name>Mapbox Terms of Service</name>
      <url>https://www.mapbox.com/legal/tos/</url>
      <distribution>repo</distribution>
    </license>
  </licenses>
  <developers>
    <developer>
      <id>mapbox</id>
      <name>Mapbox</name>
    </developer>
  </developers>
  <scm>
    <connection>scm:git@github.com:mapbox/mapbox-navigation-android.git</connection>
    <developerConnection>scm:git@github.com:mapbox/mapbox-navigation-android.git</developerConnection>
    <url>https://github.com/mapbox/mapbox-navigation-android</url>
  </scm>
</project>
```
</pre>
</details>

Reply:
Thinking about this a bit. An old me was also pushing for `api` to include these dependencies. `compileOnly` is designed for optional dependencies https://blog.gradle.org/introducing-compile-only-dependencies

https://developer.android.com/studio/build/dependencies
```
If you use this configuration, then your library module must include a runtime condition to check whether the dependency is available, and then gracefully change its behavior so it can still function if it's not provided. This helps reduce the size of the final app by not adding transient dependencies that aren't critical. This configuration behaves just like provided (which is now deprecated).
```

Maps sdk is definitely not optional for [androidauto-v0.1.0](https://github.com/mapbox/mapbox-maps-android/releases/tag/extension-androidauto-v0.1.0). I'm going back to believe this is improper use of `compileOnly`

cc: @pengdev

Reply:
Exactly, by using `compileOnly` for a required dependency, all that we're doing is hiding the version of the dependency (whether that's Maps, Nav, or Search) that Android Auto needs from downstream developers. The version they choose still has to be compatible with what the Android Auto lib was compiled against because it's required. Of course they can always upgrade (within major version) but they cannot downgrade and by obfuscating what Android Auto lib was compiled against (with usage of `compileOnly`), Gradle will not be able to figure out the correct conflict resolution by itself.

Reply:
By using `compileOnly` on Android auto extension, we are allowed to use it with earlier versions of Maps SDK from v10.0.0(even though widgets are not supported), if we use api dependency, we then have to use Maps SDK v10.4.0+, since the Android Auto widgets is built using the Maps SDK's widget feature that was introduced in v10.4.0.

At the initial release of the Android Auto extension, I think it makes sense to have a broader audience, however as we roll out more releases and v10.3.0 is outdated, it might make sense to use api dependency.

Reply:
Yeah I continue to go back and forth. But `api` dependency appears to preferred. Concluding this thread and leaving a continuation for the maps extension https://github.com/mapbox/mapbox-maps-android/issues/1341

There is also a conclusion that this could be a good way to make the search sdk an optional dependency. Moving that potential solution discussion here https://github.com/mapbox/mapbox-navigation-android/issues/5786

Reply:
Am testing the `api` dependencies downstream in 1tap with a snapshot https://github.com/mapbox/1tap-android/pull/1883

It seems, we would also need to specify specific versions to release with. Gradle doesn't really know that the latest version it is releasing with is `2.5.0-beta.3`

```
> Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
   > Could not find com.mapbox.navigation:android:2.6.0-SNAPSHOT.
     Searched in the following locations:
       - https://dl.google.com/dl/android/maven2/com/mapbox/navigation/android/2.6.0-SNAPSHOT/maven-metadata.xml
       - https://dl.google.com/dl/android/maven2/com/mapbox/navigation/android/2.6.0-SNAPSHOT/android-2.6.0-SNAPSHOT.pom
       - https://repo.maven.apache.org/maven2/com/mapbox/navigation/android/2.6.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/com/mapbox/navigation/android/2.6.0-SNAPSHOT/android-2.6.0-SNAPSHOT.pom
       - https://api.mapbox.com/downloads/v2/releases/maven/com/mapbox/navigation/android/2.6.0-SNAPSHOT/maven-metadata.xml
       - https://api.mapbox.com/downloads/v2/releases/maven/com/mapbox/navigation/android/2.6.0-SNAPSHOT/android-2.6.0-SNAPSHOT.pom
       - https://api.mapbox.com/downloads/v2/snapshots/maven/com/mapbox/navigation/android/2.6.0-SNAPSHOT/maven-metadata.xml
       - https://api.mapbox.com/downloads/v2/snapshots/maven/com/mapbox/navigation/android/2.6.0-SNAPSHOT/android-2.6.0-SNAPSHOT.pom
       - file:/root/.m2/repository/com/mapbox/navigation/android/2.6.0-SNAPSHOT/maven-metadata.xml
       - file:/root/.m2/repository/com/mapbox/navigation/android/2.6.0-SNAPSHOT/android-2.6.0-SNAPSHOT.pom
```

Reply:
So instead of depending on the project, will need to depend on specific versions